File: src/java/org/orbeon/oxf/xforms/event/XFormsEvent.java
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.xforms.event;

import org.orbeon.oxf.common.OXFException;
import org.orbeon.oxf.pipeline.StaticExternalContext;
import org.orbeon.oxf.pipeline.api.PipelineContext;
import org.orbeon.oxf.util.IndentedLogger;
import org.orbeon.oxf.xforms.XFormsConstants;
import org.orbeon.oxf.xforms.XFormsContainingDocument;
import org.orbeon.oxf.xforms.XFormsUtils;
import org.orbeon.oxf.xforms.event.events.XFormsUIEvent;
import org.orbeon.oxf.xforms.xbl.XBLContainer;
import org.orbeon.oxf.xml.XMLUtils;
import org.orbeon.oxf.xml.dom4j.LocationData;
import org.orbeon.saxon.om.*;
import org.orbeon.saxon.trans.XPathException;
import org.orbeon.saxon.value.BooleanValue;
import org.orbeon.saxon.value.SequenceExtent;
import org.orbeon.saxon.value.StringValue;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


/**
 * XFormsEvent represents an XForms event passed to all events and actions.
 */
public abstract class XFormsEvent implements Cloneable {

    private static final String XXFORMS_TYPE_ATTRIBUTE = XMLUtils.buildExplodedQName(XFormsConstants.XXFORMS_NAMESPACE_URI, "type");
    private static final String XXFORMS_TARGET_ATTRIBUTE = XMLUtils.buildExplodedQName(XFormsConstants.XXFORMS_NAMESPACE_URI, "target");
    private static final String XXFORMS_TARGETID_ATTRIBUTE = XMLUtils.buildExplodedQName(XFormsConstants.XXFORMS_NAMESPACE_URI, "targetid");
    private static final String XXFORMS_BUBBLES_ATTRIBUTE = XMLUtils.buildExplodedQName(XFormsConstants.XXFORMS_NAMESPACE_URI, "bubbles");
    private static final String XXFORMS_CANCELABLE_ATTRIBUTE = XMLUtils.buildExplodedQName(XFormsConstants.XXFORMS_NAMESPACE_URI, "cancelable");
    private static final String XXFORMS_PHASE_ATTRIBUTE = XMLUtils.buildExplodedQName(XFormsConstants.XXFORMS_NAMESPACE_URI, "phase");

    private static final String XXFORMS_REPEAT_INDEXES_ATTRIBUTE = XMLUtils.buildExplodedQName(XFormsConstants.XXFORMS_NAMESPACE_URI, "repeat-indexes");
    private static final String XXFORMS_REPEAT_ANCESTORS_ATTRIBUTE = XMLUtils.buildExplodedQName(XFormsConstants.XXFORMS_NAMESPACE_URI, "repeat-ancestors");
    private static final String XXFORMS_TARGET_PREFIXES_ATTRIBUTE = XMLUtils.buildExplodedQName(XFormsConstants.XXFORMS_NAMESPACE_URI, "target-prefixes");

    private static final String XXFORMS_EFFECTIVE_TARGETID_ATTRIBUTE = XMLUtils.buildExplodedQName(XFormsConstants.XXFORMS_NAMESPACE_URI, "effective-targetid");

    // Properties that change as the event propagates
    // TODO
//    private static final String XXFORMS_DEFAULT_PREVENTED_ATTRIBUTE = XMLUtils.buildExplodedQName(XFormsConstants.XXFORMS_NAMESPACE_URI, "default-prevented");
//    private static final String XXFORMS_CURRENT_TARGET_ATTRIBUTE = XMLUtils.buildExplodedQName(XFormsConstants.XXFORMS_NAMESPACE_URI, "current-target");// same as observer?
//    private static final String XXFORMS_OBSERVER_ATTRIBUTE = XMLUtils.buildExplodedQName(XFormsConstants.XXFORMS_NAMESPACE_URI, "observer");

    private XFormsEvent originalEvent;  // original event (for retargeted event)

    private final XBLContainer targetXBLContainer;
    private final String name;
    private XFormsEventTarget targetObject;
    private final boolean bubbles;
    private final boolean cancelable;

    public enum Phase { capture, target, bubbling }

    private Phase currentPhase = Phase.capture;

    private Map<String, SequenceExtent> customAttributes;

    // TODO: Assign for debugging
    private LocationData locationData;

    protected XFormsEvent(XFormsContainingDocument containingDocument, String name, XFormsEventTarget targetObject, boolean bubbles, boolean cancelable) {
        this.targetXBLContainer = targetObject.getXBLContainer(containingDocument);
        this.name = name;
        this.targetObject = targetObject;
        this.bubbles = bubbles;
        this.cancelable = cancelable;
    }

    public Phase getCurrentPhase() {
        return currentPhase;
    }

    public void setCurrentPhase(Phase currentPhase) {
        this.currentPhase = currentPhase;
    }

    public XBLContainer getTargetXBLContainer() {
        return targetXBLContainer;
    }
    
    protected XFormsContainingDocument getContainingDocument() {
        return targetXBLContainer.getContainingDocument();
    }

    protected IndentedLogger getIndentedLogger() {
        return getContainingDocument().getIndentedLogger(XFormsEvents.LOGGING_CATEGORY);
    }

    public String getName() {
        return name;
    }

    public XFormsEventTarget getTargetObject() {
        return targetObject;
    }

    public boolean isBubbles() {
        return bubbles;
    }

    public boolean isCancelable() {
        return cancelable;
    }

    public LocationData getLocationData() {
        return locationData;
    }

    public void setAttribute(String name, SequenceExtent value) {
        if (customAttributes == null)
            customAttributes = new HashMap<String, SequenceExtent>();
        customAttributes.put(name, value);
    }

    public SequenceIterator getAttribute(String name) {
        if ("target".equals(name) || XXFORMS_TARGET_ATTRIBUTE.equals(name) || XXFORMS_TARGETID_ATTRIBUTE.equals(name)) {// first is legacy name
            // Return the target static id

            if ("target".equals(name)) {
                getIndentedLogger().logWarning("", "event('target') is deprecated. Use event('xxforms:targetid') instead.");
            }

            return SingletonIterator.makeIterator(StringValue.makeStringValue(targetObject.getId()));
        } else if ("event".equals(name) || XXFORMS_TYPE_ATTRIBUTE.equals(name)) {// first is legacy name
            // Return the event type

            if ("event".equals(name)) {
                getIndentedLogger().logWarning("", "event('event') is deprecated. Use event('xxforms:type') instead.");
            }

            return SingletonIterator.makeIterator(StringValue.makeStringValue(this.name));
        } else if (XXFORMS_BUBBLES_ATTRIBUTE.equals(name)) {
            // Return whether the event bubbles
            return SingletonIterator.makeIterator(BooleanValue.get(bubbles));
        } else if (XXFORMS_CANCELABLE_ATTRIBUTE.equals(name)) {
            // Return whether the event is cancelable
            return SingletonIterator.makeIterator(BooleanValue.get(cancelable));
        } else if (XXFORMS_PHASE_ATTRIBUTE.equals(name)) {
            // Return the current event phase
            return SingletonIterator.makeIterator(StringValue.makeStringValue(getCurrentPhase().name()));
        } else if (XXFORMS_EFFECTIVE_TARGETID_ATTRIBUTE.equals(name)) {
            // Return the target effective id
            return SingletonIterator.makeIterator(StringValue.makeStringValue(targetObject.getEffectiveId()));
        } else if (customAttributes != null && customAttributes.get(name) != null) {
            // Return custom attribute if found
            return (customAttributes.get(name)).iterate();
        } else if ("repeat-indexes".equals(name) || XXFORMS_REPEAT_INDEXES_ATTRIBUTE.equals(name)) {

            if ("repeat-indexes".equals(name)) {
                getIndentedLogger().logWarning("", "event('repeat-indexes') is deprecated. Use event('xxforms:repeat-indexes') instead.");
            }

            final String effectiveTargetId = targetObject.getEffectiveId();
            final Integer[] parts = XFormsUtils.getEffectiveIdSuffixParts(effectiveTargetId);

            if (parts.length > 0) {
                final List<StringValue> tokens = new ArrayList<StringValue>(parts.length);
                for (final Integer currentIndex: parts) {
                    tokens.add(StringValue.makeStringValue(currentIndex.toString()));
                }
                return new ListIterator(tokens);
            } else {
                return EmptyIterator.getInstance();
            }
        } else if (XXFORMS_REPEAT_ANCESTORS_ATTRIBUTE.equals(name)) {
            // Return the target's ancestor repeat static ids if any
            final String effectiveTargetId = targetObject.getEffectiveId();
            if (XFormsUtils.hasEffectiveIdSuffix( effectiveTargetId)) {
                // There is a suffix so compute
                final List<String> ancestorRepeats
                        = getContainingDocument().getStaticState().getAncestorRepeats(XFormsUtils.getPrefixedId(targetObject.getEffectiveId()), null);

                if (ancestorRepeats.size() > 0) {
                    // At least one ancestor repeat
                    final List<StringValue> tokens = new ArrayList<StringValue>(ancestorRepeats.size());
                    for (final String currentRepeat: ancestorRepeats) {
                        // Issue: it is more correct to provide a prefixed id, but it is wrong to surface prefixed ids
                        // to form authors. The error summary, e.g., relies on unique ids, and in that case returning
                        // a prefixed id would be better. How can we solve this better?
//                        tokens.add(StringValue.makeStringValue(currentRepeat));
                        tokens.add(StringValue.makeStringValue(XFormsUtils.getStaticIdFromId(currentRepeat)));
                    }
                    return new ListIterator(tokens);
                } else {
                    // No ancestor repeats
                    return EmptyIterator.getInstance();
                }
            } else {
                // No suffix
                return EmptyIterator.getInstance();
            }
        } else if (XXFORMS_TARGET_PREFIXES_ATTRIBUTE.equals(name)) {
            // Return the target's id prefixes if any
            final String effectiveTargetId = targetObject.getEffectiveId();
            final String[] parts = XFormsUtils.getEffectiveIdPrefixParts(effectiveTargetId);

            if (parts.length > 0) {
                final List<StringValue> tokens = new ArrayList<StringValue>(parts.length);
                for (final String currentPart: parts) {
                    tokens.add(StringValue.makeStringValue(currentPart));
                }
                return new ListIterator(tokens);
            } else {
                return EmptyIterator.getInstance();
            }
        } else {
            // "If the event context information does not contain the property indicated by the string argument, then an
            // empty node-set is returned."
            getIndentedLogger().logWarning("", "Unsupported event context information for event('" + name + "').");
            return EmptyIterator.getInstance();
        }
    }

    protected void setAttributeAsString(String name, String value) {
        setAttribute(name, new SequenceExtent(new Item[] { StringValue.makeStringValue(value) } ));
    }

    protected String getAttributeAsString(String name) {
        try {
            final Item item = getAttribute(name).next();
            return (item != null) ? item.getStringValue() : null;
        } catch (XPathException e) {
            throw new OXFException(e);
        }
    }

//    protected void setAttributeAsInteger(String name, int value) {
//        setAttribute(name, new SequenceExtent(new Item[] { new IntegerValue(value) } ));
//    }

    /**
     * Attempts to get the current pipeline context using the static context.
     *
     * @return  PipelineContext, null if not found
     */
    protected PipelineContext getPipelineContext() {
        return StaticExternalContext.getStaticContext().getPipelineContext();
    }

    public XFormsEvent retarget(XFormsEventTarget newTargetObject) {
        final XFormsEvent newEvent;
        try {
            newEvent = (XFormsUIEvent) this.clone();
            newEvent.targetObject = newTargetObject;
            newEvent.originalEvent = this;
        } catch (CloneNotSupportedException e) {
            throw new OXFException(e);// should not happen because we are clonable
        }

        return newEvent;
    }

    public XFormsEvent getOriginalEvent() {
        return originalEvent != null ? originalEvent : this;
    }

    public Object clone() throws CloneNotSupportedException {
        // Default implementation is good enough
        return super.clone();
    }

    /**
     * Whether this event matches filters placed on the given event handler.
     *
     * This is used e.g. for checking modifiers and text on keypress.
     *
     * @param handler   event handler to check
     * @return          true iif the event matches the filter
     */
    public boolean matches(XFormsEventHandler handler) {
        // By default, all events match and only very specific events have special matching rules (keypress)
        return true;
    }
}

File: src/scala/org/orbeon/oxf/xforms/event/XFormsEvent.scala
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.xforms.event

import XFormsEvent._
import collection.JavaConverters._
import org.orbeon.oxf.xforms.XFormsConstants._
import org.orbeon.oxf.xforms.XFormsUtils
import org.orbeon.oxf.xforms.XFormsUtils._
import org.orbeon.oxf.xml.XMLUtils.buildExplodedQName
import org.orbeon.oxf.xml.dom4j.LocationData
import org.orbeon.saxon.om._
import org.orbeon.saxon.xqj.{SaxonXQDataFactory, StandardObjectConverter}
import org.orbeon.oxf.util.XPathCache

/**
 * Base class for all DOM events.
 *
 * - an event has a number of properties
 * - a property can be defined or not
 * - if a property is defined, it may have a value or not
 * - properties are obtained from 3 sources
 *   - XFormsEvent base properties
 *   - overriding lazyProperties
 *   - passing to the constructor
 * - the reason we use lazyProperties is that we had trouble passing a PropertyGetter with a reference to this to the
 *   constructor!
 * - all property values are Java values or Seq of Java values
 * - for Java/Scala consumers, use property[T
 * - for XPath consumers, use propertyAsIterator
 * - PropertyGetter is used instead of a plain Map because:
 *   - it allows for values to be computed dynamically, so we don't compute all property values unnecessarily
 *   - it is a PartialFunction so PropertyGetter can be easily composed with orElse
 *   - a Map is automatically a PropertyGetter
 * - whenever possible, all aspects of an event are represented as properties
 * - in some cases, events store special values natively (e.g. Throwable, ConnectionResult, etc.)
 */
abstract class XFormsEvent(
        val name: String,
        val targetObject: XFormsEventTarget,
        properties: PropertyGetter,
        val bubbles: Boolean,
        val cancelable: Boolean) {

    require(targetObject ne null)
    require(containingDocument ne null)

    final def containingDocument = targetObject.containingDocument

    // Priority is given to the base properties, then the lazyProperties, then the properties passed at construction
    private lazy val allProperties = getters(this, XFormsEvent.Getters) orElse lazyProperties orElse properties
    def lazyProperties: PropertyGetter = EmptyGetter

    // Mutable phase and observer as the event is being dispatched
    private[this] var _currentPhase: Phase = _
    private[this] var _currentObserver: XFormsEventObserver = _

    final def currentPhase: Phase = _currentPhase
    final def currentPhase_=(currentPhase: Phase): Unit = _currentPhase = currentPhase

    final def currentObserver: XFormsEventObserver = _currentObserver
    final def currentObserver_=(currentObserver: XFormsEventObserver): Unit = _currentObserver = currentObserver

    // Return a property of the given type
    // WARNING: Remember that type erasure takes place! Property[T[U1]] will work even if the underlying type was T[U2]!
    //
    // Return None if:
    // - the property is not supported
    // - it is supported but no value is available for it
    final def property[T](name: String): Option[T] =
        if (allProperties.isDefinedAt(name)) {
            // NOTE: With Scala 2.10, move to `applyOrElse`
            allProperties(name) map (_.asInstanceOf[T])
        } else {
            warnUnsupportedIfNeeded(name)
            None
        }

    // Get an attribute as an XPath SequenceIterator
    final def getAttribute(name: String): SequenceIterator = {

        warnDeprecatedIfNeeded(name)

        // NOTE: With Scala 2.10, move to `applyOrElse`
        if (allProperties.isDefinedAt(name)) {

            def propertyAsIterator = allProperties(name) map {
                case s: Seq[_] ⇒ listIterator(s map anyToItemIfNeeded)
                case other     ⇒ itemIterator(anyToItemIfNeeded(other))
            }

            propertyAsIterator getOrElse emptyIterator
        } else {
            // "If the event context information does not contain the property indicated by the string argument, then an
            // empty node-set is returned."
            warnUnsupportedIfNeeded(name)
            emptyIterator
        }
    }

    private def warnDeprecatedIfNeeded(name: String) =
        newPropertyName(name) foreach { newName ⇒
            indentedLogger.logWarning("", "event('" + name + "') is deprecated. Use event('" + newName + "') instead.")
        }

    private def warnUnsupportedIfNeeded(name: String) =
        if (warnIfMissingProperty)
            indentedLogger.logWarning("", "Unsupported event context information for event('" + name + "').")

    // These methods can be overridden by subclasses
    implicit def indentedLogger = containingDocument.getIndentedLogger(XFormsEvents.LOGGING_CATEGORY)
    def warnIfMissingProperty = true
    def newPropertyName(name: String) = Deprecated.get(name)
    def locationData: LocationData = null

    // Whether this event matches filters placed on the given event handler.
    // This is used for checking modifiers and text on keypress events.
    def matches(handler: EventHandler) = true
}

object XFormsEvent {

    // Event phases
    sealed abstract class Phase(val name: String)
    object Capture  extends Phase("capture")
    object Target   extends Phase("target")
    object Bubbling extends Phase("bubbling")

    // PropertyGetter is more general than Map, but a Map is automatically a PropertyGetter
    type PropertyGetter = PartialFunction[String, Option[Any]]
    val EmptyGetter: PropertyGetter = Map()

    // With Scala 2.10, implement applyOrElse
    def getters[E <: XFormsEvent](e: E, m: Map[String, E ⇒ Option[Any]]): PropertyGetter = new PropertyGetter {
        def isDefinedAt(name: String) = m.isDefinedAt(name)
        def apply(name: String)       = m(name)(e)
    }

    // Convert a Java object to a Saxon Item using the Saxon API
    private val anyToItem = new StandardObjectConverter(new SaxonXQDataFactory {
        def getConfiguration = XPathCache.getGlobalConfiguration
    }).convertToItem(_: Any)

    // Convert a Java object to a Saxon Item but keep unchanged if already an Item
    private val anyToItemIfNeeded: Any ⇒ Item = _ match {
        case i: Item ⇒ i
        case a ⇒ anyToItem(a)
    }

    import SingletonIterator.makeIterator

    def emptyIterator = EmptyIterator.getInstance
    def itemIterator   (i: Item)      = if (i ne null)  makeIterator(i)            else emptyIterator
    def listIterator   (s: Seq[Item]) = if (s.nonEmpty) new ListIterator(s.asJava) else emptyIterator

    def xxformsName(name: String) = buildExplodedQName(XXFORMS_NAMESPACE_URI, name)
    
    private val Deprecated = Map(
        "target"         → "xxforms:targetid",
        "event"          → "xxforms:type",
        "repeat-indexes" → "xxforms:repeat-indexes"
    )
    
    private val Getters = Map[String, XFormsEvent ⇒ Option[Any]](
        "target"                            → (e ⇒ Option(e.targetObject.getId)),
        xxformsName("target")               → (e ⇒ Option(e.targetObject.getId)),
        xxformsName("targetid")             → (e ⇒ Option(e.targetObject.getId)),
        xxformsName("absolute-targetid")    → (e ⇒ Option(XFormsUtils.effectiveIdToAbsoluteId(e.targetObject.getEffectiveId))),
        "event"                             → (e ⇒ Option(e.name)),
        xxformsName("type")                 → (e ⇒ Option(e.name)),
        xxformsName("bubbles")              → (e ⇒ Option(e.bubbles)),
        xxformsName("cancelable")           → (e ⇒ Option(e.cancelable)),
        xxformsName("phase")                → (e ⇒ Option(e.currentPhase.name)),
        xxformsName("observerid")           → (e ⇒ Option(e.currentObserver.getId)),
        xxformsName("absolute-observerid")  → (e ⇒ Option(XFormsUtils.effectiveIdToAbsoluteId(e.currentObserver.getEffectiveId))),
        "repeat-indexes"                    → repeatIndexes,
        xxformsName("repeat-indexes")       → repeatIndexes,
        xxformsName("repeat-ancestors")     → repeatAncestors,
        xxformsName("target-prefixes")      → targetPrefixes
    )

    // NOTE: should ideally be Option[Seq[Int]]. At this time XForms callers assume Option[Seq[String]].
    private def repeatIndexes(e: XFormsEvent): Option[Seq[String]] = {
        val result = Some(getEffectiveIdSuffixParts(e.targetObject.getEffectiveId).toList map (_.toString))
        result
    }

    private def repeatAncestors(e: XFormsEvent): Option[Seq[String]] =
        if (hasEffectiveIdSuffix(e.targetObject.getEffectiveId)) {
            // There is a suffix so compute
            val ancestorRepeats = e.containingDocument.getStaticOps.getAncestorRepeats(getPrefixedId(e.targetObject.getEffectiveId), null)
            Some(ancestorRepeats map getStaticIdFromId)
        } else
            // No suffix
            None

    private def targetPrefixes(e: XFormsEvent): Option[Seq[String]] =
        Some(getEffectiveIdPrefixParts(e.targetObject.getEffectiveId).toList)
}